File: tests/cpp/test_payment_plan.cpp
#include <gtest/gtest.h>
#include "engine/systems/mechanics/payment_plan.hpp"
#include "core/card_json_types.hpp"

using namespace dm::engine;
using namespace dm::core;

TEST(PaymentPlanTest, PassiveReductionSum) {
    CardDefinition def;
    def.cost = 5;

    CostReductionDef cr1;
    cr1.type = ReductionType::PASSIVE;
    cr1.reduction_amount = 2;
    cr1.id = "p1";
    cr1.name = "p1";

    CostReductionDef cr2;
    cr2.type = ReductionType::PASSIVE;
    cr2.reduction_amount = 1;
    cr2.id = "p2";
    cr2.name = "p2";

    def.cost_reductions.push_back(cr1);
    def.cost_reductions.push_back(cr2);

    PaymentPlan plan = evaluate_cost(def, 1, std::nullopt, std::nullopt);
    EXPECT_EQ(plan.total_passive_reduction, 3);
    EXPECT_EQ(plan.adjusted_after_passive, 2);
    EXPECT_EQ(plan.final_cost, 2);
}

TEST(PaymentPlanTest, ActiveReductionApplied) {
    CardDefinition def;
    def.cost = 6;

    CostReductionDef passive;
    passive.type = ReductionType::PASSIVE;
    passive.reduction_amount = 1;
    passive.id = "p1";
    passive.name = "p1";

    CostReductionDef active;
    active.type = ReductionType::ACTIVE_PAYMENT;
    active.reduction_amount = 4; // simplified prototype semantics
    active.min_mana_cost = 1;
    active.id = "a1";
    active.name = "a1";

    def.cost_reductions.push_back(passive);
    def.cost_reductions.push_back(active);

    PaymentPlan plan = evaluate_cost(def, 1, std::optional<std::string>("a1"), std::optional<int>(1));
    EXPECT_EQ(plan.total_passive_reduction, 1);
    EXPECT_EQ(plan.active_reduction_amount, 4);
    EXPECT_EQ(plan.final_cost, 1); // 6 -1 -4 =1, floor enforces >=1
}

File: tests/test_payment_plan.py
from dm_toolkit import payment


def test_evaluate_cost_passive_only():
    card = {
        'cost': 5,
        'cost_reductions': [
            {'id': 'p1', 'type': 'PASSIVE', 'amount': 2},
            {'id': 'p2', 'type': 'PASSIVE', 'reduction_per_unit': 1, 'max_units': 2},
        ]
    }
    plan = payment.evaluate_cost(card, base_cost=5, units=3)
    # p1 gives 2, p2 gives min(3,2)*1 =2 => total 4, adjusted floor 1
    assert plan.total_passive_reduction == 4
    assert plan.adjusted_after_passive == 1
    assert plan.final_cost == 1


def test_evaluate_cost_with_active():
    card = {
        'cost': 6,
        'cost_reductions': [
            {'id': 'p1', 'type': 'PASSIVE', 'amount': 1},
            {'id': 'a1', 'type': 'ACTIVE_PAYMENT', 'reduction_per_unit': 2, 'max_units': 2, 'min_mana_cost': 1},
        ]
    }
    plan = payment.evaluate_cost(card, base_cost=6, units=2, active_reduction_id='a1', active_units=2)
    # passive 1 => adjusted_after_passive 5; active reduces by 2*2=4 => final 1, with min floor 1 -> stays 1
    assert plan.total_passive_reduction == 1
    assert plan.active_reduction_amount == 4
    assert plan.final_cost == 1


def test_can_pay_with_mana_uses_plan():
    card = {'cost': 4, 'civilization': 'WATER', 'cost_reductions': [{'type': 'PASSIVE', 'amount': 1}]}
    mana_pool = {'WATER': 1, 'NATURE': 2}
    # base 4 -> passive 1 -> final 3 -> total mana 3 -> True
    assert payment.can_pay_with_mana(card, mana_pool, base_cost=4)
